stan/math/prim/prob/cauchy_lpdf.hpp

```diff
@@ -53,57 +58,76 @@ return_type_t<T_y, T_loc, T_scale> cauchy_lpdf(const T_y& y, const T_loc& mu,
-    inv_sigma[i] = 1.0 / sigma_dbl;
-    sigma_squared[i] = sigma_dbl * sigma_dbl;
-    if (include_summand<propto, T_scale>::value) {
-      log_sigma[i] = log(sigma_dbl);
-    }
-  }
+  const auto& y_arr = as_array_or_scalar(y_col);
+  const auto& mu_arr = as_array_or_scalar(mu_col);
+  const auto& sigma_arr = as_array_or_scalar(sigma_col);
 
-  for (size_t n = 0; n < N; n++) {
-    const T_partials_return y_dbl = value_of(y_vec[n]);
-    const T_partials_return mu_dbl = value_of(mu_vec[n]);
+  ref_type_t<decltype(value_of(y_arr))> y_val = value_of(y_arr);
+  ref_type_t<decltype(value_of(mu_arr))> mu_val = value_of(mu_arr);
+  ref_type_t<decltype(value_of(sigma_arr))> sigma_val = value_of(sigma_arr);
+  check_not_nan(function, "Random variable", y_val);
+  check_finite(function, "Location parameter", mu_val);
+  check_positive_finite(function, "Scale parameter", sigma_val);
 
-    const T_partials_return y_minus_mu = y_dbl - mu_dbl;
-    const T_partials_return y_minus_mu_squared = y_minus_mu * y_minus_mu;
-    const T_partials_return y_minus_mu_over_sigma = y_minus_mu * inv_sigma[n];
-    const T_partials_return y_minus_mu_over_sigma_squared
-        = y_minus_mu_over_sigma * y_minus_mu_over_sigma;
+  size_t N = max_size(y, mu, sigma);
 
-    if (include_summand<propto>::value) {
-      logp -= LOG_PI;
-    }
-    if (include_summand<propto, T_scale>::value) {
-      logp -= log_sigma[n];
-    }
-    logp -= log1p(y_minus_mu_over_sigma_squared);
+  const auto& inv_sigma
+      = to_ref_if<!is_constant_all<T_scale>::value>(inv(sigma_val));
+  const auto& y_minus_mu
+      = to_ref_if<!is_constant_all<T_y, T_loc, T_scale>::value>(y_val - mu_val);
 
-    if (!is_constant_all<T_y>::value) {
-      ops_partials.edge1_.partials_[n]
-          -= 2 * y_minus_mu / (sigma_squared[n] + y_minus_mu_squared);
-    }
-    if (!is_constant_all<T_loc>::value) {
-      ops_partials.edge2_.partials_[n]
-          += 2 * y_minus_mu / (sigma_squared[n] + y_minus_mu_squared);
+  logp -= sum(log1p(square(y_minus_mu * inv_sigma)));
+  if (include_summand<propto>::value) {
+    logp -= N * LOG_PI;
+  }
+  if (include_summand<propto, T_scale>::value) {
+    logp -= sum(log(sigma_val)) * N / size(sigma);
+  }
+
+  if (!is_constant_all<T_y, T_loc, T_scale>::value) {
+    const auto& sigma_squared
+        = to_ref_if<!is_constant_all<T_scale>::value>(square(sigma_val));
+    const auto& y_minus_mu_squared
+        = to_ref_if<!is_constant_all<T_scale>::value>(square(y_minus_mu));
+    if (!is_constant_all<T_y, T_loc>::value) {
+      const auto& mu_deriv
+          = to_ref_if < !is_constant_all<T_y>::value
+            && !is_constant_all<T_loc>::value
+                   > (2 * y_minus_mu / (sigma_squared + y_minus_mu_squared));
+      if (!is_constant_all<T_y>::value) {
+        if (is_vector<T_y>::value) {
+          ops_partials.edge1_.partials_
+              = forward_as<T_partials_array>(-mu_deriv);
```

Comment:
What does forward_as do? Why this instead of just forward? Is it casting things?

Reply:
I prefere the old name of `forward_as`, which was `assume_type`. It assumes that its input is of given type (eigen version also accepts compatible expressions). If the argument type is different it has the assumed type for return type, but it throws. This way we can within ifs with type dependant condition assume of what type a particular variable will be.

It exists just to make the code compile. Otherwise we could not assign a scalar `mu_deriv` to `ops_partials.edge1_.partials_`. But since this assignment is executed within `if (is_vector<T_y>::value) {`, we know `mu_deriv` can not be a scalar.

Reply:
Oh so it's a runtime type check + placeholder sorta thing